### ISSUE TITLE
YOLO - Prediction interpretation

### ISSUE THREAD

**Issue by Abdiflame** [Dobiasd/frugally-deep#225]
### Hello Dobiasd,

Firstly, thanks for your awesome job, is pretty helpful for Python/Keras developers to embed their models into C++ applications.

Right now, I am having some issues related to the **prediction output of my YOLO model.**
I followed this https://github.com/AntonMu/TrainYourOwnYOLO to prepare my Python YOLO model, and I managed to generate a _model+weights.h5_ file and successfully converted to _.json_ giving your _convert_model.py_ function.

My idea is just to load the model and weights on C++ and return the "bounding boxes" values from the prediction (maybe also the classes/confidence values).

However, I am facing some interpretation (maybe) problems with selecting and showing the prediction results.
Could you help me a bit?

### The code is as follow:
`
 #include <iostream>
 #include <opencv2/core/core.hpp>
 #include <opencv2/highgui/highgui.hpp>
 #include <fdeep/fdeep.hpp>
 #include <opencv2/imgproc/imgproc.hpp>

 using namespace cv;
 using namespace std;

int main()
{
	const auto model = fdeep::load_model("full_model.json");
	String image_path = "image.jpg";
	Mat dst;
	const cv::Mat image = cv::imread(image_path);
	cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
	assert(image.isContinuous());
	
	cv::resize(image, dst, cv::Size(416, 416), 0, 0, CV_INTER_NN); 
	std::cout << "Image resized" << std::endl;
	cout << "Width : " << dst.size().width << endl;
	cout << "Height: " << dst.size().height << endl;

	const auto input = fdeep::tensor_from_bytes(dst.ptr(),
		static_cast<std::size_t>(dst.rows),
		static_cast<std::size_t>(dst.cols),
		static_cast<std::size_t>(dst.channels()),
		0.0f, 1.0f);

	std::cout << "Input finished" << std::endl;
	
**// Here is where I am not sure what to do, how to select the correct prediction outputs and cout them**
	const fdeep::tensors output_tensors = model.predict({ input });
	const fdeep::tensor boxes = output_tensors[0];
	const fdeep::tensor conf = output_tensors[1];
	const fdeep::tensor label = output_tensors[2];

        std::cout << fdeep::show_tensor(boxes) << std::endl;
	std::cout << fdeep::show_tensor(conf) << std::endl;
	std::cout << "-----------------------" << std::endl;
	std::cout << fdeep::show_tensor(label) << std::endl;
	cout << "Prediction completed";
	
`
Thanks in advance.

**Comment by Dobiasd** [comment]
Hi, and thanks for the nice feedback. :)

So, if I understand correctly, the model was successfully converted to (and loaded/used in) frugally-deep, and the question is what the numbers in the output tensors mean.

They should have the same semantics as when doing the prediction in Python/TF. How do you interpret the outputs there? Or is there some code that does it for you, which you could try to understand and reproduce in C++?

**Comment by Abdiflame** [comment]
> Hi, and thanks for the nice feedback. :)
> 
> So, if I understand correctly, the model was successfully converted to (and loaded/used in) frugally-deep, and the question is what the numbers in the output tensors mean.
> 
> They should have the same semantics as when doing the prediction in Python/TF. How do you interpret the outputs there? Or is there some code that does it for you, which you could try to understand and reproduce in C++?

Hello, thanks for the fast reply.

Yes, you are right, there is a code that interprets the results. However, the code doesn't use the function model.prediction(). There is a bunch of K.cast(), K.sigmoid(), K.constant() etc. Dealing with the tensors.

The final part (where it returns the bounding box values is made by this function)

/keras_yolo3/yolo3/model.py

`def yolo_head(feats, anchors, num_classes, input_shape, calc_loss=False):
    """Convert final layer features to bounding box parameters."""
    num_anchors = len(anchors)
    # Reshape to batch, height, width, num_anchors, box_params.
    anchors_tensor = K.reshape(K.constant(anchors), [1, 1, 1, num_anchors, 2])

    grid_shape = K.shape(feats)[1:3]  # height, width
    grid_y = K.tile(
        K.reshape(K.arange(0, stop=grid_shape[0]), [-1, 1, 1, 1]),
        [1, grid_shape[1], 1, 1],
    )
    grid_x = K.tile(
        K.reshape(K.arange(0, stop=grid_shape[1]), [1, -1, 1, 1]),
        [grid_shape[0], 1, 1, 1],
    )
    grid = K.concatenate([grid_x, grid_y])
    grid = K.cast(grid, K.dtype(feats))

    feats = K.reshape(
        feats, [-1, grid_shape[0], grid_shape[1], num_anchors, num_classes + 5]
    )

    # Adjust preditions to each spatial grid point and anchor size.
    box_xy = (K.sigmoid(feats[..., :2]) + grid) / K.cast(
        grid_shape[::-1], K.dtype(feats)
    )
    box_wh = (
        K.exp(feats[..., 2:4])
        * anchors_tensor
        / K.cast(input_shape[::-1], K.dtype(feats))
    )
    box_confidence = K.sigmoid(feats[..., 4:5])
    box_class_probs = K.sigmoid(feats[..., 5:])

    if calc_loss == True:
        return grid, feats, box_xy, box_wh
    return box_xy, box_wh, box_confidence, box_class_probs`

Thanks in advance :)

**Comment by Dobiasd** [comment]
Could you collect and post the following?
- An example image with one or two known bounding boxes
- The coordinates of these boxes (output from the Python code)
- The values of the output tensors from doing the prediction in frugally-deep

Maybe if we see it that clearly, we can find the correct mapping.

**Comment by Abdiflame** [comment]
Hello Dobias, thanks again for your kind reply.
I collected the information and uploaded it in the following folder.

https://drive.google.com/file/d/1dVklh73DFCsX5ct7-XKReDyR9nKfGX4A/view?usp=sharing

Remind: The YOLO model was trained following the GitHub from AntonMu here: https://github.com/AntonMu/TrainYourOwnYOLO

The inference output from the python code is an output image (with the bounding box) and an excel file containing the inference information. 
I added the frugally tensors output from the prompt into a text file.

Best regards 👍

**Comment by Dobiasd** [comment]
Thanks for the good summary.

OK, so the output of the raw YOLO model is not some box coordinates that we just need to re-arrange in the right manner. Instead, it seems to be some kind of output image, maybe a heatmap containing the probabilities. My guess it one needs to do some kind of thresholding/segmentation on it to get actual boxes.

The Python code seems to do something like that directly in TensorFlow because the boxes are a result of `sess.run`: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo.py#L154

The part in the session generating the boxes seems to come from here: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo.py#L125

This `yolo_eval` function is defined here: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo3/model.py#L217

So I guess you'd need to somewhat re-create this logic in C++ to get the boxes, since it's not part of the YOLO model but accompanying utility code in Python.

Building this in C++ probably requires to understand what is happening in the Python code. For me this my first contact with the YOLO model, so I have no clue and would need to dig into it.

However, solving this problem does not look like a frugally-deep related issue, since fdeep is only concerned with the model itself. I suggest you could try to re-implement the box-finding code in C++ and maybe ask on StackOverflow if you get stuck at some point.

**Comment by Abdiflame** [comment]
> Thanks for the good summary.
> 
> OK, so the output of the raw YOLO model is not some box coordinates that we just need to re-arrange in the right manner. Instead, it seems to be some kind of output image, maybe a heatmap containing the probabilities. My guess it one needs to do some kind of thresholding/segmentation on it to get actual boxes.
> 
> The Python code seems to do something like that directly in TensorFlow because the boxes are a result of `sess.run`: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo.py#L154
> 
> The part in the session generating the boxes seems to come from here: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo.py#L125
> 
> This `yolo_eval` function is defined here: https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/src/keras_yolo3/yolo3/model.py#L217
> 
> So I guess you'd need to somewhat re-create this logic in C++ to get the boxes, since it's not part of the YOLO model but accompanying utility code in Python.
> 
> Building this in C++ probably requires to understand what is happening in the Python code. For me this my first contact with the YOLO model, so I have no clue and would need to dig into it.
> 
> However, solving this problem does not look like a frugally-deep related issue, since fdeep is only concerned with the model itself. I suggest you could try to re-implement the box-finding code in C++ and maybe ask on StackOverflow if you get stuck at some point.

Yes, you are right @Dobiasd.  At first, I was very confused about the show_tensor() outputs. But it seems that this problem does not directly relate to frugally-deep. As you said, it looks like a heatmap containing the probabilities from each window (anchors). 

Anyway, **really thanks** for all your kind responses 🥇 . I will keep working more on that project and hopefully decode this tensor's output successfully.

Best regards,
Cheers from Korea 👍
